cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -420,6 +419,260 @@ class self_comparator {
+        } else if (lhs_is_null != rhs_is_null) {
+          return false;
+        }
+      }
+
+      return equality_compare(lhs.element<Element>(lhs_element_index),
+                              rhs.element<Element>(rhs_element_index));
+    }
+
+    template <typename Element,
+              CUDF_ENABLE_IF(not cudf::is_equality_comparable<Element, Element>() and
+                             not cudf::is_nested<Element>())>
+    __device__ bool operator()(size_type const lhs_element_index, size_type const rhs_element_index)
+    {
+      CUDF_UNREACHABLE("Attempted to compare elements of uncomparable types.");
+    }
+
+    template <typename Element, CUDF_ENABLE_IF(cudf::is_nested<Element>())>
+    __device__ bool operator()(size_type const lhs_element_index,
+                               size_type const rhs_element_index) const noexcept
+    {
+      column_device_view lcol = lhs;
+      column_device_view rcol = rhs;
+      int l_start_off         = lhs_element_index;
+      int r_start_off         = rhs_element_index;
+      int l_end_off           = lhs_element_index + 1;
+      int r_end_off           = rhs_element_index + 1;
+      while (is_nested(lcol.type())) {
+        if (nulls) {
+          for (int i = l_start_off, j = r_start_off; i < l_end_off; ++i, ++j) {
+            bool const lhs_is_null{lcol.is_null(i)};
+            bool const rhs_is_null{rcol.is_null(j)};
+
+            if (lhs_is_null and rhs_is_null) {
+              if (nulls_are_equal == null_equality::UNEQUAL) { return false; }
+            } else if (lhs_is_null != rhs_is_null) {
+              return false;
+            }
+          }
+        }
+        if (lcol.type().id() == type_id::STRUCT) {
+          lcol = lcol.child(0);
+          rcol = rcol.child(0);
+        } else if (lcol.type().id() == type_id::LIST) {
+          auto l_list_col = detail::lists_column_device_view(lcol);
+          auto r_list_col = detail::lists_column_device_view(rcol);
+          for (int i = l_start_off, j = r_start_off; i < l_end_off; ++i, ++j) {
+            if (l_list_col.offset_at(i + 1) - l_list_col.offset_at(i) !=
+                r_list_col.offset_at(j + 1) - r_list_col.offset_at(j))
+              return false;
+          }
+          lcol        = l_list_col.child();
+          rcol        = r_list_col.child();
+          l_start_off = l_list_col.offset_at(l_start_off);
+          r_start_off = r_list_col.offset_at(r_start_off);
+          l_end_off   = l_list_col.offset_at(l_end_off);
+          r_end_off   = r_list_col.offset_at(r_end_off);
+          if (l_end_off - l_start_off != r_end_off - r_start_off) { return false; }
+        }
+      }
+
+      for (int i = l_start_off, j = r_start_off; i < l_end_off; ++i, ++j) {
+        bool equal = type_dispatcher<dispatch_void_if_nested>(
+          lcol.type(), element_comparator{nulls, lcol, rcol, nulls_are_equal}, i, j);
+        if (not equal) { return false; }
+      }
+      return true;
+    }
```

Comment:
This isn't meant to be 100% functional as it requires a few new pieces of machinery, but I wanted to convey what I think should be possible here. 

At present, my observation is that this logic is effectively zooming in on successive slices of child columns by carrying state through the `l_start_off/l_end_off` and `r_start_off/r_end_off`.

My intuition is to make this logic more natural by using the slicing mechanism inherent in `column_device_view`s and instead carry state by updating `column_device_view` objects that represent the success slices as you iterate through the nested structure 


Assumptions:
- `column_device_view::sliced_child` would just copy the parents offset to the child (doesn't modify any data/offsets) 
- `lists_column_device_view::sliced_child` would apply the parent offset plus `offset_at` to the returned slice childs offset. 


```suggestion
    {
      column_device_view left_slice = lhs;
      column_device_view right_slice = rhs;
      
      while (is_nested(lcol.type())) {
        if (nulls) {
          auto similar_nulls = thrust::equal(thrust::seq, 
                                             left_slice.null_begin(), left_slice.null_end(), 
                                             right_slice.null_begin(), 
                                             [null_are_equal](bool lhs_is_null, bool rhs_is_null){
                                                 return null_compare(lhs_is_null, rhs_is_null, nulls_are_equal);
                                             });

        if(not similar_nulls){ return false; }                                             

        if (lcol.type().id() == type_id::STRUCT) {
          left_slice  = left_slice.sliced_child(0);
          right_slice = left_slice.sliced_child(0);
        } else if (lcol.type().id() == type_id::LIST) {

          auto left_list = detail::lists_column_device_view(left_slice);
          auto right_list = detail::lists_column_device_view(right_slice);

          if(not thrust::equal(thrust::seq, 
                               left_list.begin_list_sizes(), 
                               left_list.end_list_sizes(),
                               right_list.begin_list_sizes())){
            // If the size of each sub-list isn't the same, then parent lists cannot be equal
            return false;
          }
          left_slice  = left_list.sliced_child(); 
          right_slice = right_list.sliced_child();
          // If the size of the parent lists is different, they cannot be equal
          if(left_slice.size() != right_slice.size()){
              return false;
          }
        }
      }
      auto compare = [&](auto i, auto j){
          auto comparator = element_comparator{nulls, left_slice, right_slice, nulls_are_equal};
          return type_dispatcher<dispatch_void_if_nested>{left_slice.type(), comparator, i, j};
      }
      return thrust::equal(thrust::seq, 
                           thrust::counting_iterator{0}, thrust::counting_iterator{left_slice.size()},
                           thrust::counting_iterator{0},
                           compare)
```

Reply:
In this line `column_device_view left_slice = lhs;` it should be `column_device_view left_slice = slice(lhs, {lhs_idx, lhs_idx + 1});` Because otherwise we're comparing the entire column in a single thread.

Reply:
I also have a small fear that slicing and creating new `column_device_view` objects might degrade performance. I think it doesn't right now because the compiler knows that it doesn't have to create an object and can use one directly from the `table_device_view` memory

Reply:
> I also have a small fear that slicing and creating new column_device_view objects might degrade performance.

I share that concern. No way to tell other than to try :)

Reply:
Also, I realized we can lift the `type_dispatch` up a level. There's no need to type dispatch for every element in the slice. We can just type dispatch once and compare all the elements in the slice.

Reply:
Good news. The changes had only 1% perf impact.

Reply:
Is that before or after changing the type dispatch to only be one per slice instead of per-element? I would hope that improves perf.

Reply:
7% improvement after pulling the type dispatcher out.